File: src/main/java/org/jukeboxmc/command/CommandEnum.java
package org.jukeboxmc.command;

import lombok.ToString;
import org.cloudburstmc.protocol.bedrock.data.command.CommandEnumConstraint;
import org.cloudburstmc.protocol.bedrock.data.command.CommandEnumData;

import java.util.*;

/**
 * @author Cloudburst
 * @version 1.0
 */
@ToString
public class CommandEnum {

    private final String name;
    private final List<String> values;

    public CommandEnum( String name, List<String> values ) {
        this.name = name;
        this.values = values;
    }

    public String getName() {
        return this.name;
    }

    public List<String> getValues() {
        return this.values;
    }

    public int hashCode() {
        return this.name.hashCode();
    }

    protected CommandEnumData toNetwork() {
        String[] aliases;
        if ( this.values.size() > 0 ) {
            List<String> aliasList = new ArrayList<>( this.values );
            aliasList.add( this.name );
            aliases = aliasList.toArray( new String[0] );
        } else {
            aliases = new String[]{ this.name };
        }
        return new CommandEnumData( this.name + "Aliases", this.toNetwork( aliases ), false );
    }

    protected LinkedHashMap<String, Set<CommandEnumConstraint>> toNetwork( String[] values ) {
        LinkedHashMap<String, Set<CommandEnumConstraint>> map = new LinkedHashMap<>();
        for ( String value : values ) {
            map.put( value, Collections.emptySet() );
        }
        return map;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/command/CommandEnum.kt
package org.jukeboxmc.server.command

import org.cloudburstmc.protocol.bedrock.data.command.CommandEnumConstraint
import org.cloudburstmc.protocol.bedrock.data.command.CommandEnumData

open class CommandEnum(val name: String, val values: Array<String>) {

    fun toNetwork(): CommandEnumData {
        val aliases: Array<String> = if (this.values.isNotEmpty()) {
            val list = this.values.toMutableList()
            list.add(name)
            list.toTypedArray()
        } else {
            arrayOf(this.name)
        }
        return CommandEnumData("${this.name}Aliases", toNetwork(aliases), false)
    }

    fun toNetwork(values: Array<String>): LinkedHashMap<String, Set<CommandEnumConstraint>> {
        val map = LinkedHashMap<String, Set<CommandEnumConstraint>>()
        for (value in values) {
            map[value] = emptySet()
        }
        return map
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (this.javaClass != other?.javaClass) return false
        other as CommandEnum
        if (this.name != other.name) return false
        return this.values == other.values
    }

    override fun hashCode(): Int {
        return this.name.hashCode()
    }
}

